cluster-autoscaler/cloudprovider/clusterapi/clusterapi_converters_test.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterapi
+
+import (
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+const (
+	apiVersion = "machine.openshift.io/v1beta1"
+	kind       = "MachineSet"
+	name       = "Foo"
+	namespace  = "Bar"
+)
+
+func TestConverterNewMachineSetFromUnstructured(t *testing.T) {
+	testobj := unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": apiVersion,
+			"kind":       kind,
+			"metadata": map[string]interface{}{
+				"name":      name,
+				"namespace": namespace,
+			},
+			"spec": map[string]interface{}{
+				"template": map[string]interface{}{
+					"spec": map[string]interface{}{
+						"metadata": map[string]interface{}{
+							"labels": map[string]interface{}{
+								"foo": "bar",
+							},
+						},
+						"taints": []interface{}{
+							map[string]interface{}{
+								"key":       "taintkey",
+								"value":     "taintvalue",
+								"effect":    "tainteffect",
+								"timeadded": metav1.Time{},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	machineSet := newMachineSetFromUnstructured(&testobj)
+	if machineSet.APIVersion != apiVersion {
+		t.Error("MachineSet.APIVersion improperly set")
+	}
+
+	if machineSet.Kind != kind {
+		t.Error("MachineSet.Kind improperly set")
+	}
+
+	if machineSet.Name != name {
+		t.Error("MachineSet.Name improperly set")
+	}
+
+	if machineSet.Namespace != namespace {
+		t.Error("MachineSet.Namespace improperly set")
+	}
+
+	if machineSet.Spec.Template.Spec.Labels["foo"] != "bar" {
```

Comment:
Should add a test for taints too? 

Also, might be easier to just create an expected object and compare the two, since we should be able to construct a full object `&Machine{...}` with the expected outcome, then adding future checks would be a case of updating the input and expected output `Machine`, though it would make it harder to compare when there's a failure, not sure 🤔

Reply:
i am still working on the rest of the test suite, i just wanted to get some opinions on direction before going further.

as for comparing the objects, i thought about that but as i was still building the test i just took the slow approach you see here.

Reply:
i added tests for taints, but the Time values are driving me slightly mad currently